psiminer-core/src/main/kotlin/psi/graphs/CodeGraph.kt

```diff
@@ -0,0 +1,45 @@
+package psi.graphs
+
+import com.intellij.psi.PsiElement
+import psi.graphs.edgeProviders.EdgeProvider
+import psi.preOrder
+
+class CodeGraph(val root: PsiElement) {
+    private val edges: MutableMap<PsiElement, MutableList<Edge>> = mutableMapOf()
+
+    fun getAdjacentEdges(from: PsiElement): MutableList<Edge> = edges.getOrPut(from) { mutableListOf() }
+
+    fun getAdjacentEdgesOfTypes(from: PsiElement, types: Set<EdgeType>, returnReversed: Boolean): List<Edge> =
+        getAdjacentEdges(from).filter { edge ->
+            (!edge.reversed || returnReversed) && types.contains(edge.type)
+        }.toList()
+
+    fun traverseGraph(edgeTypes: Set<EdgeType>, useReversed: Boolean, visitVertex: (PsiElement) -> Unit) {
+        val visited = mutableSetOf<PsiElement>()
+
+        fun dfs(v: PsiElement) {
+            visited.add(v)
+            visitVertex(v)
+            getAdjacentEdgesOfTypes(v, edgeTypes, useReversed).forEach { (_, to, _) ->
+                if (!visited.contains(to)) {
+                    dfs(to)
+                }
+            }
+        }
+
+        root.preOrder().forEach { v ->
+            if (!visited.contains(v)) {
+                dfs(v)
+            }
+        }
+    }
+
+    fun <T : EdgeProvider> acceptEdgeProvider(edgeProvider: T): CodeGraph {
+        val newEdges = edgeProvider.provideEdges(this)
+        newEdges.forEach { edge ->
+            getAdjacentEdges(edge.from).add(edge)
+            getAdjacentEdges(edge.to).add(edge.reversed())
```

Comment:
And I finally realized what I want here. Not DSL, but with the same mechanism under the hood.

Let's use functional interfaces to define EdgeProvider
```kotlin
fun interface EdgeProvider {
    fun accept(codeGraph: CodeGraph): CodeGraph = codeGraph.also { it.provideEdges() }
    fun CodeGraph.provideEdges(): Int
}
```

Then we can use something like that to define new edge provider:
```kotlin
val AstEdgeProvider = EdgeProvider {
    root.preOrder().flatMap { from ->
        from.children.map { to ->
                addEdge(from, to, EdgeType.AST)
            }
        }
}
```

And in the pipeline, it can be used in this way
```kotlin
val codeGraph = CodeGraph(root).apply { codeGraph ->
    edgeProviders.forEach { it.accept(codeGraph) }
}
```

Why, I think this is cool:
- CodeGraph doesn't know anything about edge providers. It just provides an interface to add new edges (`addEdge` function)
- Each edge provider is implemented as a one function as tightly as possible, easy to debug and test
- Inside edge provider, you are working with CodeGraph context -- all functions from it are available (see direct access to `root` and `addEdge`
- Functional Interface lets you create edge providers as singletons, easy use inside lists.